design/storage-class-device-set.md

```diff
@@ -0,0 +1,493 @@
+underlying storage devices. There are three main aspects to this abstraction:
+
+1. A `StorageClassDevice` is both storage and the software required to make the
+storage available and manage it in the cluster. As such, the struct takes into
+account the resources required to run the associated software, if any.
+1. A single `StorageClassDevice` could be comprised of multiple underlying
+storage devices, specified by having more than one item in the
+`volumeClaimTemplates` field.
+1. Since any storage devices that are part of a `StorageClassDevice` will be
+represented by block-mode PVCs, they will need to be associated with a Pod so
+that they can be attached to cluster nodes.
+
+A `StorageClassDeviceSet` will have the following fields associated with it:
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no
+  resource requests.
+* **placement**: The placement criteria for the devices. Default is no
+  placement criteria.
+* **config**: Granular device configuration. This is a generic
+  `map[string]string` to allow for provider-specific configuration.
+* **volumeClaimTemplates**: A list of PVC templates to use for provisioning the
+  underlying storage devices.
+
+An entry in `volumeClaimTemplates` must specify the following fields:
+  * **resources.requests.storage**: The desired capacity for the underlying
+  storage devices.
+  * **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
+
+The CephCluster CRD could be extended to include a new field:
+`spec.storage.StorageClassDeviceSets`, which would be a list of one or more
+`StorageClassDeviceSets`. If elements exist in this list, the CephCluster
+controller would then create enough PVCs to match each `StorageClassDeviceSet`'s
+`Count` field, attach them to individual OsdPrepare Jobs, then attach them to
+OSD Pods once the Jobs are completed. For the initial implementation, only one
+entry in `volumeClaimTemplates` would be supported, if only to tighten the scope
+for an MVP.
+
+The PVCs would be provisioned against a configured or default StorageClass. It
+is recommended that the admin setup a StorageClass with `volumeBindingMode:
+WaitForFirstConsumer` set.
+
+If the admin wishes to control device placement, it will be up to them to make
+sure the desired nodes are labeled properly to ensure the Kubernetes scheduler
+will distribute the OSD Pods based on Placement criteria.
+
+In keeping with current Rook-Ceph patterns, the **resources** and **placement**
+for the OSDs specified in the `StorageClassDeviceSet` would override any
+cluster-wide configurations for OSDs. Additionally, the `StorageClassDeviceSet`
+configuration would take precedence over any other conflicting configurations,
+such as if `useAllDevices` is configured for the cluster.
+
+### OSD Deployment Behavior
```

Comment:
An error case to consider is that the device set should be considered invalid for OSDs if the PV mode is not `block`.

Reply:
That would only happen if the claim template is somehow explicitly configured to `volumeNode: Filesystem`. Even then, we could always just ignore the field.